File: flutter_app/lib/strategies/ai_context.dart
import '../models/receipt_data.dart';
import 'ai_strategy.dart';

/// Context class que utilitza les estratègies d'IA.
/// Seguint el patró Strategy per a la PT5.
class AIContext {
  AIStrategy? _strategy;

  /// Estableix l'estratègia actual.
  void setStrategy(AIStrategy strategy) {
    _strategy = strategy;
  }

  /// Processa el tiquet amb l'estratègia seleccionada.
  Future<ReceiptData> process(String text) async {
    if (_strategy == null) {
      throw Exception("Cap estratègia d'IA seleccionada.");
    }
    return await _strategy!.processReceipt(text);
  }
}

File: src/AIContext.js
/**
 * Context class that uses the AI strategies
 * Following the Strategy Pattern for PT5
 */
export class AIContext {
    constructor() {
        this.strategy = null;
    }

    /**
     * Set the current strategy
     * @param {AIStrategy} strategy 
     */
    setStrategy(strategy) {
        this.strategy = strategy;
    }

    /**
     * Process receipt using the selected strategy
     * @param {string} text 
     * @returns {Promise<object>}
     */
    async process(text) {
        if (!this.strategy) {
            throw new Error("Cap estratègia d'IA seleccionada.");
        }
        return await this.strategy.processReceipt(text);
    }
}
